File: src/02.string/04.transform/03.506.go
package transform

import (
	"sort"
	"strconv"
)

func FindRelativeRanks(score []int) []string {
	desc := []string{"Gold Medal", "Silver Medal", "Bronze Medal"}
	n := len(score)
	type pair struct {
		score int
		index int
	}

	newScore := make([]pair, n)

	for i, v := range score {
		newScore[i] = pair{
			score: v,
			index: i,
		}
	}

	sort.Slice(newScore, func(i, j int) bool {
		return newScore[i].score > newScore[j].score
	})

	res := make([]string, n)

	for i, v := range newScore {
		if i <= 2 {
			res[v.index] = desc[i]
		} else {
			res[v.index] = strconv.Itoa(i + 1)
		}
	}

	return res
}

File: src/02.string/04.transform/03.506.ts
export function findRelativeRanks(score: number[]): string[] {
    /** 
     
    // 排序
    const n = score.length;
    const copyScore = [...score];
    copyScore.sort((a, b) => {
        return b - a;
    });

    const obj = {};
    for (let i = 0; i < n; i++) {
        if (i + 1 === 1) {
            obj[copyScore[i]] = "Gold Medal";
        } else if (i + 1 === 2) {
            obj[copyScore[i]] = "Silver Medal";
        } else if (i + 1 === 3) {
            obj[copyScore[i]] = "Bronze Medal";
        } else {
            obj[copyScore[i]] = String(i + 1);
        }
    }

    const res: string[] = [];
    for (let i = 0; i < n; i++) {
        res.push(obj[score[i]]);
    }

    return res;
    */

    const desc = ["Gold Medal", "Silver Medal", "Bronze Medal"];
    const n = score.length;
    const newScore: {
        score: number;
        index: number;
    }[] = [];

    for (let i = 0; i < n; i++) {
        newScore.push({
            score: score[i],
            index: i,
        });
    }

    newScore.sort((a, b) => {
        return b.score - a.score;
    });

    const res: string[] = [];
    for (let i = 0; i < n; i++) {
        if (i <= 2) {
            res[newScore[i].index] = desc[i];
        } else {
            res[newScore[i].index] = String(i + 1);
        }
    }

    return res;
}
